### PR TITLE
feat(account): Implement the Ability to Add Recovery Providers by Governance Action

### PR DESCRIPTION
**Pull request by enddynayn**
This pull request introduces new functionality to the MSA (Message Source Account) pallet by adding support for managing Recovery Providers, including their approval and removal via governance. It also updates related documentation, benchmarking, and tests to reflect these changes.

### New Features for Recovery Providers:
* Added a new storage type `RecoveryProviders` to store approved Recovery Providers.
* Introduced two new extrinsics:
  - `approve_recovery_provider`: Allows governance to approve a Recovery Provider, emitting the `RecoveryProviderApproved` event.
  - `remove_recovery_provider`: Allows governance to remove a Recovery Provider, emitting the `RecoveryProviderRemoved` event.

### Updates to Configuration and Benchmarks:
* Added `RecoveryProviderApprovalOrigin` to the configuration, defining the origin required for approving recovery providers
* Added benchmarks for the new extrinsics `approve_recovery_provider` and `remove_recovery_provider`.

### Documentation Enhancements:
* Updated the README to include details about the new extrinsics and state queries related to Recovery Providers.

### Testing Improvements:
* Added governance tests to verify the functionality of approving and removing Recovery Providers, including edge cases like unauthorized access and non-existent providers.

### Weight Adjustments:
* Updated the weight functions to include the new extrinsics `approve_recovery_provider` and `remove_recovery_provider`.

# Checklist
- [x] Updated Pallet Readme?
- [ ] Updated js/api-augment for Custom RPC APIs?
- [x] Design doc(s) updated?
- [x] Unit Tests added?
- [ ] e2e Tests added? [out of scope for this PR, tracking issue #2491]
- [x] Benchmarks added?
- [x] Spec version incremented?

### CODE REVIEW

File: pallets/msa/src/lib.rs [hunk 1]
```diff
@@ -140,6 +140,11 @@ pub mod pallet {
 		/// The origin that is allowed to create providers via governance
 		type CreateProviderViaGovernanceOrigin: EnsureOrigin<Self::RuntimeOrigin>;
 
+		/// The origin that is allowed to approve recovery providers
+		type RecoveryProviderApprovalOrigin: EnsureOrigin<
+			<Self as frame_system::Config>::RuntimeOrigin,
+		>;
```
Comment (mattheworris) [inline pallets/msa/src/lib.rs:146]: nit: match syntax for `CreateProviderApprovalOrigin`
```suggestion
		type RecoveryProviderApprovalOrigin: EnsureOrigin<Self::RuntimeOrigin>;
```

**Comment by mattheworris** [comment]
> > Does a Recovery Provider also need to be a Provider? This implies that Recovery Providers can be delegation targets; not sure if we want that.
> 
> Hmm... @JoeCap08055, You are saying that a Recovery Service could just do Recovery without being a Provider. I can see that. Here are the main reasons I think to keep them as Providers:
> 
> 1. Providers have access to Capacity. It is expected that these recovery services would use capacity
> 2. Providers (will) have additional on chain information about them published: [Design Doc: Provider Contexts #2311](https://github.com/frequency-chain/frequency/pull/2311)
> 
> I think those are the two main reasons. It also simplifies the tooling and the path. (Become a Provider, then request to be a Recovery Provider rather than: Become a provider if you want to use capacity which can happen at anytime, also become a recovery service)

What is the incentive to be a Recovery Provider, if you are not already a Provider?
I had some brief discussions with @enddynayn about the economic incentives and we may need to explore them further. 

However, it seemed that currently the only incentive to be an RP was to provide the additional recovery service to a Provider's users.

**Comment by JoeCap08055** [comment]
> > Does a Recovery Provider also need to be a Provider? This implies that Recovery Providers can be delegation targets; not sure if we want that.
> 
> Hmm... @JoeCap08055, You are saying that a Recovery Service could just do Recovery without being a Provider. I can see that. Here are the main reasons I think to keep them as Providers:
> 
> 1. Providers have access to Capacity. It is expected that these recovery services would use capacity
> 2. Providers (will) have additional on chain information about them published: [Design Doc: Provider Contexts #2311](https://github.com/frequency-chain/frequency/pull/2311)
> 
> I think those are the two main reasons. It also simplifies the tooling and the path. (Become a Provider, then request to be a Recovery Provider rather than: Become a provider if you want to use capacity which can happen at anytime, also become a recovery service)

Well, then, if being a Recovery Provider requires that you first become a Provider, we're missing some checks in this PR, I think.

Also, I thought we had an issue somewhere to track a design discussion surrounding the potential separation of the concepts of app/delegation target vs. Capacity user. I can't find the issue, but if it exists we should make a note there, and if not... do we want to create one?

**Comment by JoeCap08055** [comment]
TL;DR on preceding thread here:

* Since we don't have any near-term plans to separate the concept of "Capacity user" from "app/delegation target" (currently the conflation of these two attributes are what "being a Provider" means), a Recovery Provider *must* be a Provider (because they need to use Capacity)
* Therefore, we don't need a separate `RecoveryProviderId` type alias
* While a valid use case exists for a Provider to be ONLY a Recovery Provider, this is easily achieved by the RP simply not requesting any delegations or performing any other on-chain services within its app

**Review by mattheworris — APPROVED** [review]
- Updated branch
- Added updates to README.md
- Read through changes
- Executed `make ci-local`

**Comment by wilwade** [comment]
> Do we need an ability for a Provider to request to be a Recovery Provider? Or some other way to propose that a Provider be approved as a Recovery Provider? Or will that just be a manual governance proposal that someone creates?

I was imagining this would just be manual governance for now. Unlike a normal provider where the number would be very high, the number of recovery providers is expected to be small.

Also the "request" setup was only required for Providers as the Providers and approved via council. Recovery Providers have to submit to a full referendum

**Review by shannonwells — APPROVED** [review]
Only non-blocking comments, LGTM!

File: pallets/msa/src/lib.rs [hunk 2]
```diff
@@ -179,6 +182,13 @@ pub mod pallet {
 		OptionQuery,
 	>;
 
+	/// Storage type for approved recovery providers
+	/// - Key: Provider MSA Id
+	/// - Value: [`bool`]
+	#[pallet::storage]
+	pub type RecoveryProviders<T: Config> =
+		StorageMap<_, Twox64Concat, ProviderId, bool, OptionQuery>;
+
```
[the remark this answers is not in the document]
Comment (mattheworris) [reply pallets/msa/src/lib.rs:191]: - Added note/question to #2269

File: pallets/msa/src/lib.rs [hunk 3]
```diff
@@ -1054,10 +1074,76 @@ pub mod pallet {
...
+			ensure!(
+				Self::is_registered_provider(provider_msa_id),
+				Error::<T>::ProviderNotRegistered
+			);
+
+			RecoveryProviders::<T>::insert(ProviderId(provider_msa_id), true);
+
+			Self::deposit_event(Event::RecoveryProviderApproved {
+				provider_id: ProviderId(provider_msa_id),
+			});
```
[the remark this answers is not in the document]
Comment (mattheworris) [reply pallets/msa/src/lib.rs:1105]: - Implemented check for existing approval and updated tests to reflect the change.

File: pallets/msa/src/tests/governance_tests.rs [hunk 4]
```diff
@@ -133,3 +135,106 @@ fn propose_to_be_provider_long_name_should_fail() {
...
+			),
+			BadOrigin
+		);
+
+		assert!(Msa::is_approved_recovery_provider(&ProviderId(new_msa_id)));
+	})
+}
+
+#[test]
+fn remove_nonexistent_recovery_provider_should_succeed() {
```
[the remark this answers is not in the document]
Comment (mattheworris) [reply pallets/msa/src/tests/governance_tests.rs:229]: - Test added to ensure that no events are emitted on the 2nd approval and the state is unaffected.

**Review by JoeCap08055 — APPROVED** [review]
My questions addressed; approved.
